File: ga/asg1/graph_c/implementations/iterator.cpp
#include <vector>
#include "../interfaces/iterator.h"
#include "../interfaces/vertex.h"
#include <exception>


template <class T>
Iterator<T>::Iterator(const std::vector<T>& c): container{c} {
  this->it = this->container.begin();
}


template <class T>
void Iterator<T>::first() {
  this->it = this->container.begin();
}


template <class T>
bool Iterator<T>::valid() const {
  return this->it != this->container.end();
}



template <class T>
T Iterator<T>::getCurrent() const {
  if(this->valid())
    return *(this->it);
  else
    throw std::exception();
}


template <class T>
void Iterator<T>::next() {
  (this->it)++;
}

File: sem2/ga/asg1/graph_python/Iterator.py
from typing import List, TypeVar
from copy import deepcopy


T = TypeVar('T')


class Iterator:


    def __init__(self, list: List[T]):
        self.__c = list
        self.__idx = 0


    def first(self):
        self.idx = 0


    def valid(self) -> bool:
        return this.idx < len(self.c)


    def getCurrent(self) -> T:
        if self.valid():
            return self.c[idx]
        else:
            raise Exception


    def next(self):
        self.idx++
